app/src/editor/canvas/components/DraggableCanvasWindow.jsx

```diff
@@ -31,148 +31,141 @@ export const Title = ({ children, onClose }: BaseProps) => (
+export const Toolbar = ({ children, className, ...props }: BaseProps) => (
+    <div className={cx(styles.toolbar, className)} {...props}>
         {children}
     </div>
 )
 
 export type Layout = {
-    dragging: boolean,
-    resizing: boolean,
     x: number,
     y: number,
     width: number,
     height: number,
 }
 
-export const useLayoutState = (initialState: Layout = {}) => {
-    const [layout, setLayout] = useState({
-        dragging: false,
-        resizing: false,
-        x: 0,
-        y: 0,
-        width: 600,
-        height: 400,
-        ...initialState,
-    })
-    const updatePosition = useCallback((x: number, y: number) => {
-        setLayout((state) => ({
-            ...state,
-            x,
-            y,
-        }))
-    }, [setLayout])
-    const updateSize = useCallback((width: number, height: number) => {
-        setLayout((state) => ({
-            ...state,
-            width,
-            height,
-        }))
-    }, [setLayout])
-    const setDragging = useCallback((dragging: boolean = true) => {
-        setLayout((state) => ({
-            ...state,
-            dragging,
-        }))
-    }, [setLayout])
-    const setResizing = useCallback((resizing: boolean = true) => {
-        setLayout((state) => ({
-            ...state,
-            resizing,
-        }))
-    }, [setLayout])
-    return useMemo(() => (
-        [layout, updateSize, updatePosition, setDragging, setResizing]
-    ), [layout, updateSize, updatePosition, setDragging, setResizing])
+type CanvasWindowProps = Layout & {
+    onChangePosition: Function,
+    onChangeSize: Function,
+    children?: Node,
 }
 
-type CanvasWindowProps = {
-    position: {
-        x: number,
-        y: number,
-    },
-    size: {
-        width: number,
-        height: number,
-    },
-    onPositionUpdate: Function,
-    onSizeUpdate: Function,
-    children?: Node,
+function clamp(value, min, max) {
+    return Math.max(min, Math.min(max, value))
+}
+
+export function useLayoutState({ x = 0, y = 0, width = 600, height = 400 }: Layout = {}) {
+    // wtf flow. I don't know what it wants.
+    // $FlowFixMe
+    const [position, setPosition] = useState([x, y])
+    // $FlowFixMe
+    const [size, setSize] = useState([width, height])
+    return useMemo(() => ({
+        position,
+        setPosition,
+        size,
+        setSize,
+    }), [position, setPosition, size, setSize])
 }
 
 export const DraggableCanvasWindow = ({
-    position,
-    size,
-    onPositionUpdate,
-    onSizeUpdate,
+    x,
+    y,
+    width,
+    height,
+    onChangePosition,
+    onChangeSize,
     children,
 }: CanvasWindowProps) => {
-    const [layout, updateSize, updatePosition, setDragging, setResizing] = useLayoutState()
+    const { position, setPosition, size, setSize } = useLayoutState({
+        x,
+        y,
+        width,
+        height,
+    })
+
+    // $FlowFixMe
+    const [dragging, setDragging] = useState(false)
+    // $FlowFixMe
```

Comment:
These `$FlowFixMe`s aren't necessary, it seems.

Reply:
Ah you're right. Removed.